src/libponyc/pass/docgen.c

```diff
@@ -755,6 +798,179 @@ static void doc_methods(docgen_t* docgen, docgen_opt_t* docgen_opt,
     doc_method(docgen, docgen_opt, p->ast);
 }
 
+static char* concat(const char *s1, const char *s2)
+{
+  size_t str_size = strlen(s1) + strlen(s2) + 1;
+  char* result = (char*) malloc(str_size); //+1 for the null-terminator
+  strcpy(result, s1);
+  strcat(result, s2);
+  result[str_size - 1] = '\0';
+  return result;
+}
+
+static doc_sources_t* copy_source_to_doc_src(docgen_t* docgen, source_t* source, const char* package_name)
+{
+  pony_assert(docgen != NULL);
+  pony_assert(source != NULL);
+  pony_assert(package_name != NULL);
+
+  doc_sources_t* result = (doc_sources_t*) calloc(sizeof(doc_sources_t), 1);
+
+  char filename_copy[FILENAME_MAX];
+  strcpy(filename_copy, source->file);
+
+  const char* filename = get_file_name(filename_copy);
+  const char* filename_without_ext = remove_ext(filename, '.', 0);
+  const char* filename_md_extension = concat(filename_without_ext, ".md");
+
+  // Absolute path where a copy of the source will be put.
+  char source_dir[FILENAME_MAX];
+  path_cat(docgen->doc_source_dir, package_name, source_dir);
+
+  //Create directory for [documentationDir]/src/[package_name]
+  pony_mkdir(source_dir);
+  
+  // Get absolute path for [documentationDir]/src/[package_name]/[filename].md
+  char* path = (char*) malloc(sizeof(char) * FILENAME_MAX);
+  path_cat(source_dir, filename_md_extension, path);
+
+  // Get relative path for [documentationDir]/src/[package_name]/
+  // so it can be written in the mkdocs.yml file.
+  const char* doc_source_dir_relative = concat("src/", package_name);
+  doc_source_dir_relative = concat(doc_source_dir_relative, "/");
+
+  // Get relative path for [documentationDir]/src/[package_name]/[filename].md
+  const char* doc_path = concat(doc_source_dir_relative, filename_md_extension);
+
+  // Add a number (starting a 0) at the end of the file 
+  // in case there is name clashing.
+  // Only go up to 10 before not trying anymore.
+  bool are_name_clashing = false;
+  int current_numbering = 0;
+  int number_of_try = 0;
+  do {
+    are_name_clashing = false;
+    for (size_t i = 0; i < docgen->included_sources_count; i++) {
+      pony_assert(docgen->included_sources[i] != NULL);
+      // Check if the path we want to use was used for another source file.
+      if(strcmp(path, docgen->included_sources[i]->file_path) == 0) {
+        are_name_clashing = true;
+      }
+    }
+    number_of_try++;
+    // If the name are clashing add --[number] before extension to try
+    // to avoid name clashes.
+    if (are_name_clashing) {
+      current_numbering++;
+      char extension[FILENAME_MAX];
+      sprintf(extension, "--%i.md", current_numbering);
+      free((void*) filename_md_extension);
+      filename_md_extension = concat(filename_without_ext, extension);
+      path_cat(source_dir, filename_md_extension, path);
+      free((void*) doc_path);
+      doc_path = concat(doc_source_dir_relative, filename_md_extension);
```

Comment:
Are you sure? From my POV it nees to be destroyed directly if we don't manage to create the `.md` file for the code. Otherwise it needs to be destroyed once the documentation is generated (which I forget to do, it seems).